File: src/typescript/src/pipeline_services/core/runtime_pipeline.ts
import {
  Action,
  Pipeline,
  PipelineResult,
  StepAction,
} from "./pipeline.js";
import { RemoteSpec, http_step } from "../remote/http_step.js";

type RuntimeAction<ContextType> =
  | Action<ContextType>
  | StepAction<ContextType>
  | RemoteSpec;

function asAction<ContextType>(
  action: RuntimeAction<ContextType>,
): Action<ContextType> | StepAction<ContextType> {
  if (action instanceof RemoteSpec) {
    return (context: ContextType) =>
      http_step(action, context) as Promise<ContextType>;
  }
  if (typeof action !== "function") {
    throw new TypeError("Action must be callable or a RemoteSpec");
  }
  return action;
}

/** @deprecated Construct a Pipeline directly. */
export class RuntimePipeline<ContextType = unknown> {
  private currentValue: ContextType | null;
  private ended = false;
  private readonly preActions: Array<
    Action<ContextType> | StepAction<ContextType>
  > = [];
  private readonly actions: Array<
    Action<ContextType> | StepAction<ContextType>
  > = [];
  private readonly postActions: Array<
    Action<ContextType> | StepAction<ContextType>
  > = [];
  public lastResult: PipelineResult<ContextType> | null = null;

  constructor(
    public readonly name: string,
    public readonly shortCircuitOnException: boolean = true,
    initial: ContextType | null = null,
  ) {
    this.currentValue = initial;
  }

  value(): ContextType | null {
    return this.currentValue;
  }

  reset(value: ContextType): void {
    this.currentValue = value;
    this.ended = false;
    this.lastResult = null;
  }

  clearRecorded(): void {
    this.preActions.length = 0;
    this.actions.length = 0;
    this.postActions.length = 0;
  }

  async addPreAction(
    action: RuntimeAction<ContextType>,
  ): Promise<ContextType | null> {
    return this.addAndExecute("preActions", this.preActions, action);
  }

  async addAction(
    action: RuntimeAction<ContextType>,
  ): Promise<ContextType | null> {
    return this.addAndExecute("actions", this.actions, action);
  }

  async addPostAction(
    action: RuntimeAction<ContextType>,
  ): Promise<ContextType | null> {
    return this.addAndExecute("postActions", this.postActions, action);
  }

  async add_pre_action(
    action: RuntimeAction<ContextType>,
  ): Promise<ContextType | null> {
    return this.addPreAction(action);
  }

  async add_action(
    action: RuntimeAction<ContextType>,
  ): Promise<ContextType | null> {
    return this.addAction(action);
  }

  async add_post_action(
    action: RuntimeAction<ContextType>,
  ): Promise<ContextType | null> {
    return this.addPostAction(action);
  }

  toImmutable(): Pipeline<ContextType> {
    const pipeline = new Pipeline<ContextType>(
      this.name,
      this.shortCircuitOnException,
    );
    for (const action of this.preActions) {
      pipeline.addPreAction(action);
    }
    for (const action of this.actions) {
      pipeline.addAction(action);
    }
    for (const action of this.postActions) {
      pipeline.addPostAction(action);
    }
    return pipeline.freeze();
  }

  to_immutable(): Pipeline<ContextType> {
    return this.toImmutable();
  }

  freeze(): Pipeline<ContextType> {
    return this.toImmutable();
  }

  private async addAndExecute(
    phase: string,
    destination: Array<
      Action<ContextType> | StepAction<ContextType>
    >,
    action: RuntimeAction<ContextType>,
  ): Promise<ContextType | null> {
    if (this.ended) {
      return this.currentValue;
    }
    if (this.currentValue == null) {
      throw new TypeError(
        "RuntimePipeline requires a non-null current value",
      );
    }

    const normalized = asAction(action);
    destination.push(normalized);
    const pipeline = new Pipeline<ContextType>(
      `${this.name}:runtime`,
      this.shortCircuitOnException,
    );
    if (phase === "preActions") {
      pipeline.addPreAction(normalized);
    } else if (phase === "postActions") {
      pipeline.addPostAction(normalized);
    } else {
      pipeline.addAction(normalized);
    }

    const result = await pipeline.runDetailed(this.currentValue);
    this.currentValue = result.context;
    this.lastResult = result;
    this.ended = result.shortCircuited;
    return this.currentValue;
  }
}

File: src/Python/pipeline_services/core/runtime_pipeline.py
from __future__ import annotations

import warnings
from typing import Generic, TypeVar

from .pipeline import Action, Pipeline, PipelineResult, StepAction
from ..remote.http_step import RemoteSpec, http_step

ContextType = TypeVar("ContextType")
RuntimeAction = Action[ContextType] | StepAction[ContextType] | RemoteSpec


def _as_action(action: RuntimeAction[ContextType]) -> Action[ContextType] | StepAction[ContextType]:
    if isinstance(action, RemoteSpec):
        return lambda context: http_step(action, context)
    if not callable(action):
        raise TypeError("Action must be callable or a RemoteSpec")
    return action


class RuntimePipeline(Generic[ContextType]):
    """Deprecated immediate-execution helper backed by the canonical Pipeline runner."""

    def __init__(
        self,
        name: str,
        short_circuit_on_exception: bool = True,
        current: ContextType | None = None,
    ) -> None:
        warnings.warn(
            "RuntimePipeline is deprecated; construct a Pipeline directly.",
            DeprecationWarning,
            stacklevel=2,
        )
        self.name = name
        self.short_circuit_on_exception = short_circuit_on_exception
        self.current = current
        self.ended = False
        self.pre_actions: list[Action[ContextType] | StepAction[ContextType]] = []
        self.actions: list[Action[ContextType] | StepAction[ContextType]] = []
        self.post_actions: list[Action[ContextType] | StepAction[ContextType]] = []
        self.last_result: PipelineResult[ContextType] | None = None

    def value(self) -> ContextType | None:
        return self.current

    def reset(self, value: ContextType) -> None:
        self.current = value
        self.ended = False
        self.last_result = None

    def clear_recorded(self) -> None:
        self.pre_actions.clear()
        self.actions.clear()
        self.post_actions.clear()

    def recorded_pre_action_count(self) -> int:
        return len(self.pre_actions)

    def recorded_action_count(self) -> int:
        return len(self.actions)

    def recorded_post_action_count(self) -> int:
        return len(self.post_actions)

    def add_pre_action(self, action: RuntimeAction[ContextType]) -> ContextType | None:
        return self._add_and_execute("preActions", self.pre_actions, action)

    def add_action(self, action: RuntimeAction[ContextType]) -> ContextType | None:
        return self._add_and_execute("actions", self.actions, action)

    def add_post_action(self, action: RuntimeAction[ContextType]) -> ContextType | None:
        return self._add_and_execute("postActions", self.post_actions, action)

    def freeze(self) -> Pipeline[ContextType]:
        return self.to_immutable()

    def to_immutable(self) -> Pipeline[ContextType]:
        pipeline = Pipeline[ContextType](self.name, self.short_circuit_on_exception)
        for action in self.pre_actions:
            pipeline.add_pre_action(action)
        for action in self.actions:
            pipeline.add_action(action)
        for action in self.post_actions:
            pipeline.add_post_action(action)
        return pipeline.freeze()

    def _add_and_execute(
        self,
        phase: str,
        destination: list[Action[ContextType] | StepAction[ContextType]],
        action: RuntimeAction[ContextType],
    ) -> ContextType | None:
        if self.ended:
            return self.current
        if self.current is None:
            raise ValueError("RuntimePipeline requires a non-None current value")

        normalized = _as_action(action)
        destination.append(normalized)

        pipeline = Pipeline[ContextType](
            f"{self.name}:runtime",
            self.short_circuit_on_exception,
        )
        if phase == "preActions":
            pipeline.add_pre_action(normalized)
        elif phase == "postActions":
            pipeline.add_post_action(normalized)
        else:
            pipeline.add_action(normalized)

        result = pipeline.run_detailed(self.current)
        self.current = result.context
        self.last_result = result
        self.ended = result.short_circuited
        return self.current
